Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,504 @@
+
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            switch self {
+            case .http1(let provider):
+                return provider.getConnection(preference: preference)
+            case .future(let futureProvider):
+                return futureProvider.flatMap { provider in
+                    provider.getConnection(preference: preference)
+                }
+            }
+        }
+
+        func release(connection: Connection) {
+            switch self {
+            case .http1(let provider):
+                return provider.release(connection: connection)
+            case .future(let futureProvider):
+                futureProvider.whenSuccess { provider in
+                    provider.release(connection: connection)
+                }
+            }
+        }
+
+        func syncClose() throws {
+            switch self {
+            case .http1(let provider):
+                try provider.syncClose()
+            case .future(let futureProvider):
+                try futureProvider.wait().syncClose()
+            }
+        }
+    }
+
+    class HTTP1ConnectionProvider {
+        let eventLoop: EventLoop
+        let configuration: HTTPClient.Configuration
+        let key: ConnectionPool.Key
+        private var state: State
+        var isClosed: NIOAtomic<Bool>
+        let maximumConcurrentConnections: Int = 8
+        let lock = Lock()
+        let parentPool: ConnectionPool
+
+        init(group: EventLoopGroup, key: ConnectionPool.Key, configuration: HTTPClient.Configuration, initialConnection: Connection, parentPool: ConnectionPool) {
+            self.eventLoop = initialConnection.channel.eventLoop
+            self.configuration = configuration
+            self.key = key
+            self.parentPool = parentPool
+            self.isClosed = NIOAtomic.makeAtomic(value: false)
+            self.state = State(initialConnection: initialConnection)
+            self.configureCloseCallback(of: initialConnection)
+        }
+
+        deinit {
+            // FIXME: Add the checks back
+            /*
+             assert(availableConnections.isEmpty, "Available connections should be empty before deinit")
+             assert(leased == 0, "All leased connections should have been returned before deinit")
+             */
+        }
+
+        private func resolvePreference(_ preference: HTTPClient.EventLoopPreference) -> (EventLoop, Bool) {
+            switch preference.preference {
+            case .indifferent:
+                return (self.eventLoop, false)
+            case .delegate(let el):
+                return (el, false)
+            case .delegateAndChannel(let el), .testOnly_exact(let el, _):
+                return (el, true)
+            }
+        }
+
+        /// Get a new connection from this provider
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
```

Comment:
I think this function needs a rule, which is that the returned future must always belong to the event loop of the `Connection` being returned. This is mostly do-able, but I think the `.leaseConnection` clause doesn't do that. I'd also like this contract called out in a doc comment.

Reply:
@Lukasa  This hasn't been fixed yet, by _mostly doable_ do you mean we should do it as best as we can with no guarantee that it will always be the case? Because I think it can't _always_ be the case.

Reply:
CC @Lukasa

Reply:
In some sense it doesn't matter _which_ event loop this future returns on, so long as it is always consistent about that choice. Otherwise it becomes impossible to write thread-safe programs. Given that we allow users to express an event loop preference here it seems sensible to follow the idea that the event loop should be returned on their preference.

There are two good choices of which event loop the future callback should run on:

1. The event loop of the returned connection.
2. The event loop of the caller.

Both of these are defensible, depending on what the user is exactly trying to do. It doesn't matter too much which of those we choose, but we should be attempting to guarantee that this is the case. Here we lose track of what the event loop of the future we're returning is, so it'd be good to keep track of it.

Reply:
@Lukasa Now the returned future respects the preference of the user, note that if the preference is set to `.indifferent`, the future can be *any* on any EL, just like the channel.